Mqtt cover tilt calculated from position

m1ch opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
So far it is not possible to tilt mqtt-cover with Venetian blinds. This are generally blinds, that tilt and change the position at the same time. 

If the `tilt_from_position` is set to true, tilt is activated in the frontend and can be used as usual, even without a `tilt_command_topic` set. 
This includes:
* An update of the tilt, when the position changes. 
* A change of the position when the tilt is set.

The position is calculated:
tilt = old_tilt/100*tilt_max + (old_position - new_postion)/tilt_max*100)
if tilt < 0: tilt = 0
if tilt > 100: tilt = 100

To set the tilt it adds or subtracts the given value to the position. 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
cover:
  - platform: mqtt
    name: "Zimmer1 Rollo"
    command_topic:      'shellies/blind-zimmer1_1/roller/0/command'
    state_topic:        'shellies/blind-zimmer1_1/roller/0' 
    position_topic:     'shellies/blind-zimmer1_1/roller/0/pos'
    # tilt_status_topic must not be used
    set_position_topic: 'shellies/blind-zimmer1_1/roller/0/command/pos'
    # tilt_command_topic must not be used
    position_open: 100
    position_closed: 0
    tilt_min: 0
    tilt_max: 6 # Percent it takes the cover to turn arround
    tilt_from_position: True # new option
    tilt_opened_value: 3
    tilt_closed_value: 0
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/16986

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

m1ch committed:
Merge pull request #1 from home-assistant/dev

sync

m1ch committed:
Merge branch 'dev' of github.com:home-assistant/core into dev

m1ch committed:
Implemented a feature to soft tile venecian blinds

homeassistant commented:
Hi @m1ch, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#47860).

Once you do that we will be able to review and accept this pull request.

Thanks!

emontnemery commented:
Can you explain a bit more how this feature is intended to be used?
When `tilt_from_position` is enabled, tilt is calculated from the cover's position?
How is it calculated?

It is possible to set tilt, and when doing so, the cover will be opened or closed to the position where the right level of tilt is reached?

emontnemery commented on homeassistant/components/mqtt/cover.py:
```diff
@@ -446,7 +482,10 @@ def supported_features(self):
-        if self._config.get(CONF_TILT_COMMAND_TOPIC) is not None:
+        if (
+            self._config.get(CONF_TILT_COMMAND_TOPIC) is not None
+            or self._config.get(CONF_TILT_STATE_SOFT) is not None
```
```suggestion
            or self._config.get(CONF_TILT_STATE_SOFT)
```

m1ch commented:
> Can you explain a bit more how this feature is intended to be used?
> When `tilt_from_position` is enabled, tilt is calculated from the cover's position?
> How is it calculated?
> 
> It is possible to set tilt, and when doing so, the cover will be opened or closed to the position where the right level of tilt is reached?

I updated the escription and added a link to the PR for the docu.

m1ch committed:
Update homeassistant/components/mqtt/cover.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

m1ch committed:
run black

m1ch committed:
Update CONF_ variable name to match option

m1ch committed:
Add appriviation

emontnemery commented:
@m1ch There are several ways in which tilt can be combined with position.
What you propose here is maybe something similar to this: https://www.youtube.com/watch?v=9uHv634jnVw ?

I think this is a bit too specific to add as a feature to MQTT cover, @thecode suggested it would be possible to support something more flexible by templating.

Would you mind finishing off PR https://github.com/home-assistant/core/pull/45751 instead, that would then give all templates full access to the entity's attributes and state.

m1ch commented:
Hi @emontnemery,

I am trying to solve this problem now for quite some time now. 

My last attempt before the change I did was this post on reddit:
https://www.reddit.com/r/homeassistant/comments/m2bvf6/set_static_variable_in_a_template/

If you now a possibility to get the old possition-value, or  store a static variable from within a template, than it would be easy. Otherwise I would still suggest to accept this PR as I found the problem to pop up quite some times during my research. Also the impact on performance for existing users is minimal. 

By the way: In the Video you send me there is a man controlling the blind with a remote control. I don't know how that can help me, as I am looking for a way to tild my blinds from an automation. I switched over from FHEM and Homematic and there the automation was working fine.It's the only thing I coudn't solve in a stable satisfacting way since I switched over from FHEM a couple of years back.

m1ch commented:
@emontnemery, 
> Would you mind finishing off PR #45751 instead, that would then give all templates full access to the entity's attributes and state.
The two PR don't overlap in the code. Merging should be easy. Once I implemented the test cases for my PR I will have a look at our PR. The idea is nice!

emontnemery commented:
> By the way: In the Video you send me there is a man controlling the blind with a remote control. But it's basically the kind of blind I am using. 

Sorry if I was being unclear, I just wanted to understand if this PR is intended to allow controlling tilt and position of the kind of blind in the video, which has a hysteresis in the movement when changing directions.
There are also other configurations, where the blind first closes fully, and if the closing continues the tilt is also closed.

emontnemery commented:
> Once I implemented the test cases for my PR I will have a look at our PR

I would suggest to first explore if controlling your blinds can be achieved by templates.
Basically, what you need to do is to pass in `entity_id=self.entity_id` when rendering templates, for example here:
https://github.com/home-assistant/core/blob/5174f63fd85fb752ce378a5c0615b2f035695918/homeassistant/components/mqtt/cover.py#L539-L546

You should then be able to implement the desired functionality by accessing the state template functions in templates: https://www.home-assistant.io/docs/configuration/templating/#states

thecode commented:
> @emontnemery,
> 
> > Would you mind finishing off PR #45751 instead, that would then give all templates full access to the entity's attributes and state.
> > The two PR don't overlap in the code. Merging should be easy. Once I implemented the test cases for my PR I will have a look at our PR. The idea is nice!

@m1ch I agree that your PR does not overlap the other one, but if we can have a solution that will work with templates it will be better for dealing with different types of covers. My first idea was to store the previous location and pass it to the template, but if we pass the entity to the template it can be used to get the current position from the template. It will also allow using other attributes of the entity, For example, when you want to first check a Boolean state if the entity is on and return zero when it is off.
If we can solve your request by extending templates it would be much better IMO.
Do you think it will work? We can work on it together if you want to finish it faster.

m1ch commented:
@thecode; @emontnemery Thank you for your inputs and I agree that a solution via templates might be more flexible. 

I have the following thoughts about that: 
It is required to know these three parameters for the calculation of the tilt: 
* The new position - this will be the payload from MQTT
* The previous tilt state - this is available via 
* The previous position state - here it gets tricky. If async_set_cover_positio and async_set_cover_tilt_positio are used with the same MQTT topic as trigger, it is not possible to determine which function is called first. A possible workaround for that would be to set position and tilt_position via one of the two functions and don't do anything in the second, if the topics are equal. This is similar to my current implementation, but that templates are possible for tilt and position. 

A short list of the benefits of an implementation via template: 
+ No additional parameters for the configuration are necessary. 
+ Possible for a better calculation of the tilt via template
+ More flexibility to control the tilt via template
- Bigger impact on the code
- The configuration will get complex (see my example on reddit) 

Please let me know what you think. The implementation itself should be straight forward. (especially once the entity is available in the templates ;) )


One additional thought to my implementation: It only uses the HA internal state-value in percent to determine the tilt. 
It uses max_tilt to define how many percent it takes the blind from one max to the other. In my opinion this works for any blind that tilts when it changes the direction.

emontnemery commented:
@m1ch I think that's a good summary, although I'm not sure about the code impact which you list as a con.
The code change should be limited to this, plus the same changes for the command topics: https://github.com/home-assistant/core/pull/45751/files#diff-abb5c6d9aaba927346669899ce35bda8b11fef1d9778bcafe5c36ae67930439c

Fully agreed that the configuration will get complex though.

For "the previous position state" - is that a blocking issue if the templates only have access to the state and attributes, or do you mean it makes the configuration trickier?

m1ch commented:
@emontnemery I am afraid it is. 

The calculation for the tilt is:
new_tilt = old_tilt - movement
movement = new_position - old_position 

And it is necessary to know if the async function for tilt, or position is called first, since it makes a difference if the position state is already the new one.

emontnemery commented:
> I think here the cover is a bit special, as it subscribes to two topics for position and tilt and if the same topic is used for both subscriptions, it is not easy to determine which function is called first.

OK, but isn't it the same problem in this PR?

m1ch commented:
> OK, but isn't it the same problem in this PR?
No. In the implementation for this pull request it is not necessary to subscribe to an extra MQTT topic for the tilt. The tilt is updated with the position. I can't think of a way to do this with templates without more impact on the code.

m1ch commented:
@emontnemery I just did a quick "sketch" how an implementation via templates could be done and here is what I sought of: 
https://github.com/m1ch/home-assistant/commit/5c47fdb81b0274d71f3ef954aec0fb154ae92ece#diff-abb5c6d9aaba927346669899ce35bda8b11fef1d9778bcafe5c36ae67930439c 

This is just a sketch for visualization and not the implementation. 

Once  #45751 is done, it will be even simpler. Please give me your feedback if you prefer the change to be something like this sketch.

thecode commented:
@m1ch looking at the sketch it looks like the only change is adding `self._position` to the render:
```python
payload = tilt_status_template.async_render_with_possible_json_value(
                payload, 
                self._position
            )
 ```
 would it also work in case #45751 to get this from the entity by entity_id?
 In my opinion the change in your sketch is better than what is currently in the PR, since it allows the template to set the final behavior of the tilt, so code parts like (just an example):
 ```python
   if self._position > 50:
                        self._tilt_value = 100
  ```
  Would not have to be hardcoded or externalized by parameters, instead they would be in the template which will allow every user to modify them.
I think it would be also OK to add your template as an example in the documentation for the type of blinds you are using.

m1ch commented:
@thecode 

> @m1ch looking at the sketch it looks like the only change is adding `self._position` to the render:

No. I also do:
* If tilt_status_topic == position_topic: update tilt in the function to update the position
* don't call the function for update tilt if tilt_status_topic == position_topic


> 
> ```python
> payload = tilt_status_template.async_render_with_possible_json_value(
>                 payload, 
>                 self._position
>             )
> ```
> 
> would it also work in case #45751 to get this from the entity by entity_id?

This part would work with entity_id. This was also my comment, that the it would be simpler with #45751 implemented. 


> In my opinion the change in your sketch is better than what is currently in the PR, since it allows the template to set the final behavior of the tilt, so code parts like (just an example):
> 
> ```python
>   if self._position > 50:
>                        self._tilt_value = 100
> ```
> 
> Would not have to be hardcoded or externalized by parameters, instead they would be in the template which will allow every user to modify them.
> I think it would be also OK to add your template as an example in the documentation for the type of blinds you are using.

emontnemery commented:
How about allowing `position_template` to return a dict `{"position": <>, "tilt_position": <>}` together with #45751?

m1ch commented:
@emontnemery 

> How about allowing `position_template` to return a dict `{"position": <>, "tilt_position": <>}` together with #45751?

This would be perfect!

emontnemery commented:
@m1ch sounds good! would you be OK to implement #45751 (you can limit it to MQTT cover if it's too much work) + open a new PR or refactor this one to allow `position_template` to return a dict?

To allow `position_template` to return a dict, maybe split up `tilt_message_received`:
```diff
         @callback
         @log_messages(self.hass, self.entity_id)
         def tilt_message_received(msg):
             """Handle tilt updates."""
-            payload = msg.payload
+            _tilt_message_received(msg.payload)
+
+        def _tilt_message_received(payload):
            tilt_status_template = self._config.get(CONF_TILT_STATUS_TEMPLATE)
            if tilt_status_template is not None:
```

And something like this in `position_message_received`:
```diff
             if template is not None:
                 payload = template.async_render_with_possible_json_value(payload)
+
+                if isinstance(payload, dict):
+                    _tilt_message_received(payload["tilt_value"])
+                    payload = payload["position"]
```

m1ch commented:
@emontnemery I start the work at #45751 now.

thecode commented:
@m1ch if you need help in implementation, unit tests or just testing the PR, feel free ask me, we can also discuss on discord.

m1ch commented:
New PR #48625 created

emontnemery commented:
@m1ch should this be closed in favor of #48625 ?

m1ch commented:
@emontnemery Please close this PR

m1ch commented:
Found the button
